cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+    UNIT_MASK = (USED_BITS_PER_UNIT >= (8U * sizeof(uint32_t))) ? 0xFFFFFFFF : (0x01U << USED_BITS_PER_UNIT) - 1,
+    /// This is the bit-mask for each item
+    ITEM_MASK = (BITS_PER_ITEM >= (8U * sizeof(uint32_t))) ? 0xFFFFFFFF : (0x01U << BITS_PER_ITEM) - 1
+  };
+
+  //------------------------------------------------------------------------------
+  // ACCESSORS
+  //------------------------------------------------------------------------------
+public:
+  __device__ __forceinline__ uint32_t Get(uint32_t index) const
+  {
+    int32_t target_offset = index * BITS_PER_ITEM;
+    uint32_t val          = 0;
+
+#pragma unroll
+    for (int32_t i = 0; i < NUM_TOTAL_UNITS; ++i)
+    {
+      // In case the bit-offset of the counter at <index> is larger than the bit range of the current unit,
+      // the bit_shift amount will be larger than the bits provided by this unit.
+      // If the bit-offset of the counter at <index> is smaller than the bit range of the current unit,
+      // the computed bit-offset will be negative, which, once casted to an unsigned type will be larger than the bits
+      // provided by this unit.
+      uint32_t bit_shift = static_cast<uint32_t>(target_offset - i * USED_BITS_PER_UNIT);
+      val |= (data[i] >> bit_shift) & ITEM_MASK;
+    }
+    return val;
+  }
+
+  __device__ __forceinline__ void Add(uint32_t index, uint32_t value)
+  {
+    int32_t target_offset = index * BITS_PER_ITEM;
+
+#pragma unroll
+    for (int32_t i = 0; i < NUM_TOTAL_UNITS; ++i)
+    {
+      // In case the bit-offset of the counter at <index> is larger than the bit range of the current unit,
+      // the bit_shift amount will be larger than the bits provided by this unit.
+      // If the bit-offset of the counter at <index> is smaller than the bit range of the current unit,
+      // the computed bit-offset will be negative, which, once casted to an unsigned type will be larger than the bits
+      // provided by this unit.
+      uint32_t bit_shift = static_cast<uint32_t>(target_offset - i * USED_BITS_PER_UNIT);
+      data[i] += (value << bit_shift) & UNIT_MASK;
+    }
+  }
+
+  __device__ __forceinline__ BitPackedCounter operator+(const BitPackedCounter &rhs) const
+  {
+    BitPackedCounter result;
+    for (int32_t i = 0; i < NUM_TOTAL_UNITS; ++i)
+    {
+      result.data[i] = data[i] + rhs.data[i];
+    }
+    return result;
+  }
+
+  //------------------------------------------------------------------------------
+  // CONSTRUCTORS
+  //------------------------------------------------------------------------------
+  __device__ __forceinline__ BitPackedCounter()
+  {
+#pragma unroll
+    for (int32_t i = 0; i < NUM_TOTAL_UNITS; ++i)
+    {
+      data[i] = 0;
+    }
+  }
+
+  //------------------------------------------------------------------------------
+  // MEMBER VARIABLES
+  //------------------------------------------------------------------------------
+private:
+  BackingUnitT data[NUM_TOTAL_UNITS];
```

Comment:
Could we `brace-init` this here and default the default constructor?
```suggestion
  BackingUnitT data[NUM_TOTAL_UNITS] = {};
```

The good question would be whether this would be properly unrolled? @dkolsen-pgi could you give me a hint here?